### ISSUE TITLE
Hook/event after store restored but before accessed?

### ISSUE THREAD

**Issue by lougreenwood** [simplabs/ember-simple-auth#1643]
Following on from #990 I'm attempting to refactor my code to directly use the session for persisting data (and avoid direct access to `store.persist`/`restore`).

In every new session I need to:
* Take a copy of a session value from the previous session
* Overwrite that value with a new unique value for this session

I need this to occur after the store has been restored into session, but before session is available to be accessed (so the overwritten value is available and I don't re-use old values from previous sessions).

It seems that `session.init()` would be a good place for this, something simple like:
```
init() {
    this._super(...arguments)

    // Store previous thing for use in places which require history of last session's thing
    this.set('previousThing', this.thing)
    this.set('thing', this._generateThing())
},
```

However, it seems that the session hasn't been restored and isn't available at this point in time.

Is there any clean & suggested route to achieve this? I've manage to hack around the problem in my implementation, but the result is quite fragile and still needs to use `session.store.restore()`.

**Comment by marcoow** [comment]
This is the perfect use case for the [`restore` method of the authenticator](http://ember-simple-auth.com/api/classes/BaseAuthenticator.html#method_restore). That method gets called with the session data that has been restored from the session store and is supposed to return a promise that resolves with new session data (if the old session does indeed constitute a valid session - if not, the returned promise should reject and the session will not be authenticated after that).

**Comment by marcoow** [comment]
`restore` is called when the application starts or whenever the session data changes. I'd say if you need these values in the authenticator when you're authenticating the session, the easiest way is probably to not store them in the session at all then but directly in `localStorage` or even a service if they don't need to be persisted.

**Comment by lougreenwood** [comment]
Maybe it's a consequence of EphemeralStore (or maybe I'm just completely wrong...?), but this twiddle seems to suggest that `restore` doesn't get called on application start-up (or at all in this case)...

https://ember-twiddle.com/69bc40b4eb6a4460d9903635df02d522

**Comment by lougreenwood** [comment]
I've done some more digging, it seems that `internal-session.restore()` (https://github.com/simplabs/ember-simple-auth/blob/master/addon/internal-session.js) does indeed only work when the session is authenticated:

* `authenticatoryFactory` will be undefined if not authenticated (https://github.com/simplabs/ember-simple-auth/blob/master/addon/internal-session.js#L68)
  * hardcoding this will allow the code to proceed but...
* `delete restoredContent.authenticated.authenticator;` silently fails if empty

So to summarise...

If I hard code my `authenticatorFactory` and wrap `delete restoredContent.authenticated.authenticator;` in an if, I can trigger the `authenticator.restore()` method to run on application startup *when not authenticated*.

It seems that this is the intended behaviour when considering that docs say "This method is invoked by the session either on application startup if session data is restored from the session store" - so I presume this behaviour is a bug...

I'd try my hand at a PR, but it seems that the multi-authenticator design of ESA makes this too complex for a fix for a noob like me - for example, how can ESA know which authenticator to run on application startup IF there are multiple authenticators? It seems that this "bug" works around this by only restoring when authenticated and the authenticator can be known...

Perhaps on app startup all authenticate restore methods could be run? But that seems like a bad idea (side effects & un-intended consequences)


Perhaps the simplest way for me to work around this is to use my own initialiser to run restore for my specific custom authenticator on app startup...

**Comment by marcoow** [comment]
> Maybe it's a consequence of EphemeralStore (or maybe I'm just completely wrong...?), but this twiddle seems to suggest that restore doesn't get called on application start-up (or at all in this case)...

When you're using the EphemeralStore there can be no session restore at all as everything is lost once you reload (since the store is ephemeral).

Tbh, I don't really understand what you're trying to do but here's some general info which maybe helps to clarify this:

* when the app starts up and the session for that app instance is created, that session tries to restore session data from the session store
  * if there is session data that contains an authenticator, the session lets that authenticator restore a session from that data (as only the authenticator can know what that data actually _"means"_ as every authenticator uses different data)
  * if the authenticator decides the session data constitutes a valid session, the session will be authenticated, otherwise the session data is cleared and the session is unauthenticated
  * if there is no session data, the session remains unauthenticated anyway
* ESA supports multiple authenticators but the session can only be authenticated with **one authenticator at a time** - consider the case where you allow people to login with credentials, or Facebook or Twitter. Once a user logs in with either of these options, there's no point really in logging in again with another one of those means as they are authenticated in the context of your app already

**Comment by lougreenwood** [comment]
Thanks for the reply. First, let me try to explain my goal...

My app uses OAuth 2.0 Authorisation code with PKCE  for authorisation: 
https://www.oauth.com/oauth2-servers/pkce/

Practically, this means that when a user attempts to login, the app must generate a `codeVerifier` which is used to generate a `codeChallenge`. The user is then redirected off site to the OAuth2 login page. After the user enters valid login creds, the user is redirected back to the app, where the app must re-send the previously generated `codeVerifier` back to the OAuth server to get it's auth tokens.

There's a good description of the flow here:
https://developer.okta.com/authentication-guide/implementing-authentication/auth-code-pkce/

Practically this means that I need to store the generated `codeVerifier` somewhere which persists between 2 sessions (because of the redirect off-site to the OAuth login page mid flow).

I opted to use the `ember-simple-auth` session store, for this, since this seems a sensible place to store auth-related data.

so... the problem I have is that if the user is not authenticated the session store data is never restored. I have worked around this by using a custom initialiser to restore the session data for a specific authenticator.

The problem seems to be that because a session is not authenticated, it is not possible for `ESA` to infer which authenticator to restore (even if there is only 1 authenticator registered in the host app).

I understand the technical constraint here, but the point of my questions has been to clarify between:
* Do the docs need clarifying so it is obvious that the session store is only restored if there is an authenticated session - this is the current description: 
** > Restores the session from a session data object. **This method is invoked by the session either on application startup if session data is restored from the session store** or when properties in the store change due to external events (e.g. in another tab) and the new session data needs to be validated for whether it constitutes an authenticated session.
** http://ember-simple-auth.com/api/classes/BaseAuthenticator.html#method_restore
* Is this actually a bug in the intended functionality (which the above description implies, assuming the above description is accurate.

To be clear, I have worked around this for my use case by creating a custom initialiser which is able to restore the store for a specific authenticator. Obviously I would prefer to remove that initialiser and fall back to the default `ESA` behaviour if `ESA` were able to restore session store a when not authenticated.
